### PR TITLE
Add probability distribution transformation APIs 

### PR DESCRIPTION
**Pull request by cxxly**
### PR types

New features
### PR changes

APIs
### Describe

Adds 13 transformation APIs and 2 distribution APIs :

new transformation APIs:

1. Transform
2. AbsTransform
3. AffineTransform
4. ChainTransform
5. ExpTransform
6. IndependentTransform
7. PowerTransform
8. ReshapeTransform
9. SigmoidTransform
10. SoftmaxTransform
11. StackTransform
12. StickBreakingTransform
13. TanhTransform

new distribution APIs:

1. Independent
2. TransformedDistribution
 

Examples:

```python
import paddle

x = paddle.to_tensor([1., 2.])
affine = paddle.distribution.AffineTransform(paddle.to_tensor(0.), paddle.to_tensor(1.))

print(affine.forward(x))
# Tensor(shape=[2], dtype=float32, place=Place(gpu:0), stop_gradient=True,
#        [1., 2.])
print(affine.inverse(affine.forward(x)))
# Tensor(shape=[2], dtype=float32, place=Place(gpu:0), stop_gradient=True,
#        [1., 2.])
print(affine.forward_log_det_jacobian(x))
# Tensor(shape=[1], dtype=float32, place=Place(gpu:0), stop_gradient=True,
#        [0.])

```

### CODE REVIEW

File: python/paddle/distribution/transform.py [hunk 1]
```diff
@@ -0,0 +1,1255 @@
...
+    'TanhTransform'
+]
+
+
+class Type(enum.Enum):
+    """Mapping type of a transformation.
+    """
+    BIJECTION = 'bijection'  # bijective(injective and surjective)
+    INJECTION = 'injection'  # injective-only
+    SURJECTION = 'surjection'  # surjective-inly
```
Comment (iclementine) [inline python/paddle/distribution/transform.py]: typo

**Review by iclementine — DISMISSED** [review]

**Review by TCChenlong — DISMISSED** [review]
LGTM
TODO：Add Chinese documentation

File: python/paddle/distribution/transform.py (hunk 1 above)
Comment (cxxly) [reply python/paddle/distribution/transform.py]: updated

File: python/paddle/distribution/transform.py [hunk 2]
```diff
@@ -0,0 +1,1255 @@
...
+
+            import paddle
+
+            x = paddle.to_tensor([1., 2.])
+            affine = paddle.distribution.AffineTransform(paddle.to_tensor(0.), paddle.to_tensor(1.))
+
+            print(affine.forward(x))
+            # Tensor(shape=[2], dtype=float32, place=Place(gpu:0), stop_gradient=True,
+            #        [1., 2.])
+            print(affine.inverse(power.forward(x)))
```
Comment (jeff41404) [inline python/paddle/distribution/transform.py]: "power" means paddle.distribution.PowerTransform？

**Comment by jeff41404** [comment]
Examples in Describe above: print(affine.inverse(power.forward(x))), "power" means paddle.distribution.PowerTransform? or should be affine?

File: python/paddle/distribution/transform.py (hunk 2 above)
Comment (cxxly) [reply python/paddle/distribution/transform.py]: 已更新，粘贴错误，粘贴了一个旧的代码示例

**Comment by cxxly** [comment]
> ```python
> import paddle
> 
> x = paddle.to_tensor([1., 2.])
> affine = paddle.distribution.AffineTransform(paddle.to_tensor(0.), paddle.to_tensor(1.))
> 
> print(affine.forward(x))
> # Tensor(shape=[2], dtype=float32, place=Place(gpu:0), stop_gradient=True,
> #        [1., 2.])
> print(affine.inverse(affine.forward(x)))
> # Tensor(shape=[2], dtype=float32, place=Place(gpu:0), stop_gradient=True,
> #        [1., 2.])
> print(affine.forward_log_det_jacobian(x))
> # Tensor(shape=[1], dtype=float32, place=Place(gpu:0), stop_gradient=True,
> #        [0.])
> ```



> Examples in Describe above: print(affine.inverse(power.forward(x))), "power" means paddle.distribution.PowerTransform? or should be affine?

是 ''affine''，粘贴错误，文档和PR描述均已更新

File: python/paddle/distribution/distribution.py [hunk 3]
```diff
@@ -96,7 +96,7 @@ def prob(self, value):
         Args:
             value (Tensor): value which will be evaluated
         """
-        raise NotImplementedError
+        return self.log_prob(value).exp()
```
Comment (jeff41404) [inline python/paddle/distribution/distribution.py:114]: shall we add some abstract methods in base class, e.g. mean, variance and rsample

Comment (cxxly) [reply python/paddle/distribution/distribution.py:114]: 已添加，这样能保持不同子类持有的方法一致；存量 normal, uniform, categorical目前是``NotImplementedError``，按照设计文档计划后续统一更新

File: python/paddle/distribution/__init__.py [hunk 4]
```diff
@@ -12,15 +12,20 @@
...
+from paddle.distribution.beta import Beta
+from paddle.distribution.categorical import Categorical
+from paddle.distribution.dirichlet import Dirichlet
+from paddle.distribution.distribution import Distribution
+from paddle.distribution.exponential_family import ExponentialFamily
+from paddle.distribution.independent import Independent
+from paddle.distribution.kl import kl_divergence, register_kl
+from paddle.distribution.multinomial import Multinomial
+from paddle.distribution.normal import Normal
+from paddle.distribution.transform import *  # noqa: F403
```
Comment (XiaoguangHu01) [inline python/paddle/distribution/__init__.py:25]: 这里需要import *的原因是什么呢？

**Review by jeff41404 — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

File: python/paddle/distribution/__init__.py (hunk 4 above)
Comment (cxxly) [reply python/paddle/distribution/__init__.py:25]: ``distribution/transform.py`` 文件中定义了需要公开API的__ALL__列表，在``distribution/__init__``中用``import *`` 全部导出，并添加到``__init__.py`` 的__all__列表中， 可以通过``paddle.distribution.xxx``访问，访问路径和竞品保持一致

**Pull request merged**